File: threads/src/main/java/im/threads/business/transport/Transport.java
package im.threads.business.transport;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.Lifecycle;

import java.util.List;

import im.threads.business.chat_updates.ChatUpdateProcessor;
import im.threads.business.logger.core.LoggerEdna;
import im.threads.business.models.ConsultInfo;
import im.threads.business.models.Survey;
import im.threads.business.models.UserPhrase;
import im.threads.business.rest.queries.BackendApi;
import im.threads.business.rest.queries.ThreadsApi;
import io.reactivex.Completable;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

public abstract class Transport {
    private final ChatUpdateProcessor chatUpdateProcessor = ChatUpdateProcessor.getInstance();
    private CompositeDisposable compositeDisposable;

    public void markMessagesAsRead(List<String> uuidList) {
        LoggerEdna.info(ThreadsApi.REST_TAG, "markMessagesAsRead : " + uuidList);
        subscribe(
                Completable.fromAction(() -> BackendApi.get().markMessageAsRead(uuidList).execute())
                        .subscribeOn(Schedulers.io())
                        .subscribe(
                                () -> {
                                    LoggerEdna.info(
                                            ThreadsApi.REST_TAG, "messagesAreRead : " + uuidList
                                    );
                                    for (String messageId : uuidList) {
                                        chatUpdateProcessor.postIncomingMessageWasRead(messageId);
                                    }
                                },
                                e -> {
                                    LoggerEdna.info(
                                            ThreadsApi.REST_TAG, "error on messages read : " + uuidList
                                    );
                                    chatUpdateProcessor.postError(new TransportException(e.getMessage()));
                                }
                        )

        );
    }

    private boolean subscribe(final Disposable event) {
        if (compositeDisposable == null || compositeDisposable.isDisposed()) {
            compositeDisposable = new CompositeDisposable();
        }
        return compositeDisposable.add(event);
    }


    public abstract void init();

    public abstract void sendRatingDone(Survey survey);

    public abstract void sendResolveThread(boolean approveResolve);

    public abstract void sendUserTying(String input);

    public abstract void sendInit();

    /**
     * TODO THREADS-6292: this method can potentially lead to messages stuck in STATE_SENDING
     */
    public abstract void sendMessage(UserPhrase userPhrase, ConsultInfo consultInfo, final String filePath, final String quoteFilePath);

    public abstract void sendClientOffline(String clientId);

    public abstract void updateLocation(Double latitude, Double longitude);

    @NonNull
    public abstract String getToken() throws TransportException;

    public abstract void setLifecycle(@Nullable Lifecycle lifecycle);
}

File: threads/src/main/java/im/threads/business/transport/Transport.kt
package im.threads.business.transport

import androidx.lifecycle.Lifecycle
import im.threads.business.chat_updates.ChatUpdateProcessor
import im.threads.business.logger.LoggerEdna.info
import im.threads.business.models.ConsultInfo
import im.threads.business.models.Survey
import im.threads.business.models.UserPhrase
import im.threads.business.rest.queries.BackendApi.Companion.get
import im.threads.business.rest.queries.ThreadsApi
import im.threads.business.serviceLocator.core.inject
import io.reactivex.Completable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers

/**
 * Предоставляет базовые методы для получения и отправки данных через веб-сокет
 */
abstract class Transport {
    private val chatUpdateProcessor: ChatUpdateProcessor by inject()
    private var compositeDisposable: CompositeDisposable? = null

    fun markMessagesAsRead(uuidList: List<String?>) {
        info(ThreadsApi.REST_TAG, "markMessagesAsRead : $uuidList")
        subscribe(
            Completable.fromAction { get().markMessageAsRead(uuidList)?.execute() }
                .subscribeOn(Schedulers.io())
                .subscribe(
                    {
                        info(
                            ThreadsApi.REST_TAG,
                            "messagesAreRead : $uuidList"
                        )
                        for (messageId in uuidList) {
                            if (messageId != null) {
                                chatUpdateProcessor.postIncomingMessageWasRead(messageId)
                            }
                        }
                    }
                ) { e: Throwable ->
                    info(
                        ThreadsApi.REST_TAG,
                        "error on messages read : $uuidList"
                    )
                    chatUpdateProcessor.postError(TransportException(e.message))
                }
        )
    }

    private fun subscribe(event: Disposable): Boolean {
        if (compositeDisposable == null || compositeDisposable?.isDisposed == true) {
            compositeDisposable = CompositeDisposable()
        }
        return compositeDisposable?.add(event) == true
    }

    abstract fun init()
    abstract fun sendRatingDone(survey: Survey)
    abstract fun sendResolveThread(approveResolve: Boolean)
    abstract fun sendUserTying(input: String)
    abstract fun sendInit()
    internal abstract fun updatePushToken()

    /**
     * TODO THREADS-6292: this method can potentially lead to messages stuck in STATE_SENDING
     */
    abstract fun sendMessage(userPhrase: UserPhrase, consultInfo: ConsultInfo?, filePath: String?, quoteFilePath: String?)
    abstract fun sendClientOffline(clientId: String)
    abstract fun updateLocation(latitude: Double, longitude: Double)
    abstract fun getToken(): String
    abstract fun setLifecycle(lifecycle: Lifecycle?)
}
